Intermittent: failing $()/@$() subprocess form does not raise CalledProcessError under load

### Summary

Under load, a failing captured (`$(...)`) or inject (`@$(...)`) subprocess
form intermittently does **not** raise `CalledProcessError`, even with
`$XONSH_SUBPROC_RAISE_ERROR` enabled. For a standalone `@$(...)` it then
surfaces as `XonshError: subprocess mode: command is empty`.

xonsh 0.23.7, Linux.

### Observed

Running the full `pytest` suite while building the package, the **same
revision** intermittently fails on loaded build hosts — seen on both
`aarch64` and `x86_64`. It passes when the test runs in isolation, so it
is timing-dependent:

```
    def test_inject_form_raises(xonsh_execer, raise_env):
        with pytest.raises(CalledProcessError):
>           xonsh_execer.exec("@$(ls /__nope__)\n")

xonsh/procs/specs.py:796: in build
    raise xt.XonshError("xonsh: subprocess mode: command is empty")
E   xonsh.tools.XonshError: xonsh: subprocess mode: command is empty
```

### Analysis

`_check_subproc_helper_raise` (`xonsh/built_ins.py`) decides whether to
raise:

```python
rtn = getattr(cp, "returncode", None)
if rtn is None or rtn == 0:
    return
```

This conflates "return code **undetermined**" (`None`) with "succeeded"
(`0`). `CommandPipeline.returncode` is a `blocking_property` that calls
`self.end()`, but `end()` is a no-op once `self.ended` is set (it already
ran in `_run_specs`), so it returns `self.proc.returncode` as-is — which
can still be `None` if the underlying proc has not been reaped yet.

When that happens the helper silently skips the raise; the inject helper
then returns an empty token list and the outer command is built empty.

### Suggested direction

The raise-check (or `CommandPipeline.returncode`) should not treat an
undetermined return code as success — it should obtain a definitive code
before deciding. Filing for maintainer input rather than a PR: reliable
reproduction needs a loaded machine and the fix touches the concurrent
pipeline code.

--- a/CLAUDE.md
+++ b/CLAUDE.md
@@ -130,7 +130,7 @@ The project follows consistent, well-organized patterns throughout the codebase 
 
 **Style:** flat, procedural — tests are standalone functions, never grouped into classes. Use `pytest.mark.parametrize` for variations.
 
-**LLM-generated tests:** when Claude creates a substantial block of tests for a feature or module, place them in a `test_<topic>_llm.py` file (e.g. `tests/parsers/test_parser_fstring_llm.py`). This keeps generated tests separate from hand-written ones.
+**LLM-generated tests:** when Claude creates a substantial block of tests for a module, place them in a `test_<module>_llm.py` file whose name mirrors the module under test — e.g. tests for `xonsh/procs/pipelines.py` go in `tests/procs/test_pipelines_llm.py`, tests for `xonsh/tools.py` go in `tests/test_tools_llm.py`. Do *not* name the file after the specific feature or bug being covered (`test_returncode_llm.py`, `test_fix_6456_llm.py`): module-mirroring keeps related generated tests in one place and prevents a long tail of one-off files. Topic-specific names are acceptable only when the surface under test really is feature-scoped and spans multiple modules (e.g. `tests/parsers/test_parser_fstring_llm.py`). This keeps generated tests separate from hand-written ones while staying easy to locate from the module path.
 
 **Key fixtures** (from `xonsh/pytest/plugin.py`):
 - `xession` — mocked xonsh session (most commonly used)

--- a/tests/procs/test_pipelines_llm.py
+++ b/tests/procs/test_pipelines_llm.py
@@ -12,9 +12,7 @@ def test_returncode_finalizes_unreaped_proc(xonsh_session):
     definite int — not ``None`` — so the raise-check helpers don't
     conflate "undetermined" with "succeeded".
     """
-    pipeline: CommandPipeline = xonsh_session.execer.eval(
-        "!(/bin/sh -c 'exit 7')"
-    )
+    pipeline: CommandPipeline = xonsh_session.execer.eval("!(/bin/sh -c 'exit 7')")
     pipeline.end()
     real_rc = pipeline.proc.returncode
     assert real_rc == 7  # sanity

--- a/tests/procs/test_pipelines_llm.py
+++ b/tests/procs/test_pipelines_llm.py
@@ -1,0 +1,80 @@
+"""LLM-generated tests for ``xonsh.procs.pipelines``."""
+
+from xonsh.procs.pipelines import CommandPipeline
+from xonsh.pytest.tools import skip_if_on_windows
+
+
+@skip_if_on_windows
+def test_returncode_finalizes_unreaped_proc(xonsh_session):
+    """Regression for #6456: if ``tee_stdout`` bails before reaping
+    (intermittent under load), ``end()`` leaves ``proc.returncode``
+    unset.  ``CommandPipeline.returncode`` must still produce a
+    definite int — not ``None`` — so the raise-check helpers don't
+    conflate "undetermined" with "succeeded".
+    """
+    pipeline: CommandPipeline = xonsh_session.execer.eval(
+        "!(/bin/sh -c 'exit 7')"
+    )
+    pipeline.end()
+    real_rc = pipeline.proc.returncode
+    assert real_rc == 7  # sanity
+
+    class _UnreapedProc:
+        returncode = None
+
+        def poll(self):
+            self.returncode = real_rc
+            return real_rc
+
+    pipeline.proc = _UnreapedProc()
+    assert pipeline.returncode == real_rc
+
+
+@skip_if_on_windows
+def test_returncode_stays_none_when_proc_still_alive(xonsh_session):
+    """The poll-based fallback added for #6456 must not block: a
+    still-alive proc keeps ``returncode`` at ``None`` so suspended /
+    background pipelines still report "undetermined" to the
+    raise-checks (which intentionally skip on ``None``).
+    """
+    pipeline: CommandPipeline = xonsh_session.execer.eval("!(true)")
+    pipeline.end()
+    assert pipeline.proc.returncode == 0
+
+    class _AliveProc:
+        returncode = None
+
+        def poll(self):
+            return None  # proc still running
+
+    pipeline.proc = _AliveProc()
+    assert pipeline.returncode is None
+
+
+@skip_if_on_windows
+def test_returncode_poll_failure_trace(xonsh_session, monkeypatch, capsys):
+    """When ``$XONSH_SUBPROC_TRACE`` is enabled, a failing ``poll()`` in
+    the #6456 fallback must surface its exception on stderr instead of
+    silently swallowing it — so that the rare "proc disappeared" path
+    is debuggable.  With trace off, the swallow stays silent.
+    """
+    pipeline: CommandPipeline = xonsh_session.execer.eval("!(true)")
+    pipeline.end()
+
+    class _PollFails:
+        returncode = None
+
+        def poll(self):
+            raise RuntimeError("simulated poll failure")
+
+    pipeline.proc = _PollFails()
+
+    # Trace off — silent swallow, returncode is None.
+    monkeypatch.setitem(xonsh_session.env, "XONSH_SUBPROC_TRACE", 0)
+    assert pipeline.returncode is None
+    assert "simulated poll failure" not in capsys.readouterr().err
+
+    # Trace on — exception surfaces on stderr; returncode is still None.
+    monkeypatch.setitem(xonsh_session.env, "XONSH_SUBPROC_TRACE", 1)
+    assert pipeline.returncode is None
+    assert "simulated poll failure" in capsys.readouterr().err

--- a/tests/procs/test_pipelines_llm.py
+++ b/tests/procs/test_pipelines_llm.py
@@ -11,13 +11,18 @@ def test_returncode_finalizes_unreaped_proc(xonsh_session):
     unset.  ``CommandPipeline.returncode`` must still produce a
     definite int — not ``None`` — so the raise-check helpers don't
     conflate "undetermined" with "succeeded".
+
+    Use a successful ``!(true)`` as the carrier pipeline (so the test
+    is independent of ``$XONSH_SUBPROC_CMD_RAISE_ERROR``, which the
+    in-CI ``run-tests.xsh`` enables) and inject the bug state by
+    swapping ``pipeline.proc`` for one whose ``returncode`` is ``None``
+    until ``poll()`` is called.
     """
-    pipeline: CommandPipeline = xonsh_session.execer.eval(
-        "!(/bin/sh -c 'exit 7')"
-    )
+    pipeline: CommandPipeline = xonsh_session.execer.eval("!(true)")
     pipeline.end()
-    real_rc = pipeline.proc.returncode
-    assert real_rc == 7  # sanity
+    assert pipeline.proc.returncode == 0  # sanity
+
+    real_rc = 7
 
     class _UnreapedProc:
         returncode = None

--- a/xonsh/procs/pipelines.py
+++ b/xonsh/procs/pipelines.py
@@ -1019,11 +1019,36 @@ class CommandPipeline:
 
     @blocking_property
     def returncode(self):
-        """Process return code, waits until command is completed."""
+        """Process return code.
+
+        ``None`` means the proc has not exited (typically a suspended or
+        backgrounded pipeline).  After a foreground pipeline has
+        ``end()``-ed this never returns ``None``.
+
+        Background: under load, ``tee_stdout`` can bail before the
+        trailing ``proc.wait()`` in ``iterraw`` runs (e.g. an uncaught
+        ``OSError`` while draining the pipe).  ``end()`` then flips
+        ``self.ended = True`` in its ``finally`` with
+        ``self.proc.returncode`` still unset.  The raise-check helpers
+        conflated ``None`` with ``0`` (success) and silently swallowed
+        real failures (#6456).  Reap once more via ``poll()`` —
+        non-blocking, so suspended/backgrounded pipelines still
+        correctly report ``None``.
+        """
         self.end()
         if self.proc is None:
             return 1
-        return self.proc.returncode
+        rc = self.proc.returncode
+        if rc is None:
+            try:
+                rc = self.proc.poll()
+            except Exception:
+                if XSH.env.get("XONSH_SUBPROC_TRACE", False):
+                    xt.print_exception(
+                        source_msg="Trace: returncode poll() fallback failed"
+                    )
+                rc = None
+        return rc
 
     @property
     def args(self):

